### PR TITLE
fix: Always keep FunctionLikeStorage->param_lookup in sync

### LINKED ISSUE #6050: FunctionLikeAnalyzer: params & param_lookup are not always symetrical
I stumbled upon the following error in the context of `psalm-plugin-laravel` development (see https://github.com/psalm/psalm-plugin-laravel/pull/160):

```
In FunctionLikeAnalyzer.php line 819:
                                                                                                                                                                                                         
  Uncaught ErrorException: Undefined array key 2 in /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php:819                          
  Stack trace:                                                                                                                                                                                           
  #0 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php(819): Laravel\Lumen\Application->Laravel\Lumen\Concerns\{closure}()         
  #1 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php(607): Psalm\Internal\Analyzer\FunctionLikeAnalyzer->checkParamReferences()  
  #2 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php(1779): Psalm\Internal\Analyzer\FunctionLikeAnalyzer->analyze()                     
  #3 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php(1447): Psalm\Internal\Analyzer\ClassAnalyzer->analyzeClassMethod()                 
  #4 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php(416): Psalm\Internal\Analyzer\ClassAnalyzer->analyzeTraitUse()                     
  #5 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(214): Psalm\Internal\Analyzer\ClassAnalyzer->analyze()                              
  #6 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(348): Psalm\Internal\Analyzer\FileAnalyzer->analyze()                                   
  #7 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(609): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}()             
  #8 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(277): Psalm\Internal\Codebase\Analyzer->doAnalysis()                                    
  #9 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(642): Psalm\Internal\Codebase\Analyzer->analyzeFiles()                           
  #10 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/src/Psalm/Internal/Cli/Psalm.php(361): Psalm\Internal\Analyzer\ProjectAnalyzer->check()                                         
  #11 /path/to/psalm-plugin-laravel/vendor/vimeo/psalm/psalm(4): Psalm\Internal\Cli\Psalm::run()                                                                                       
  #12 {main}                                                                                                                                                                                             
    thrown                                                                                                                                                                                                                                                                                                                                                                                  
```

This is the corresponding code location:

https://github.com/vimeo/psalm/blob/bee4e83ecab35409377a9591e1fe268e463b1303/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php#L814-L822

Here's what the variable state looked like at the crash:

```php
$var_name = '$state';
$storage->param_lookup = ['parameters', 'count', state']
$position = 2
$storage->params = [
  0 => ['name' => 'count', /* ... */],
  1 => ['name' => 'state', /* ... */],
]
```

So basically `param_lookup` considered an additional parameter, coming from the parent implementation, while `params` only had the current implementation's parameters.

The causing code was this stub for Laravel:

```php
<?php

namespace Illuminate\Database\Eloquent\Factories;

trait HasFactory
{
    /**
     * @template TCountOrState of callable|array|int|null
     *
     * @param  TCountOrState  $count
     * @param  callable|array  $state
     *
     * @return (
        TCountOrState is int
            ? FactoryBuilder<static, TCountOrState>
            : FactoryBuilder<static, 1>
        )
     */
    public static function factory($count = null, $state = [])
    {

    }
}
```

The real `HasFactory` trait in Laravel 8 looks like this though:

```php
<?php

namespace Illuminate\Database\Eloquent\Factories;

trait HasFactory
{
    /**
     * Get a new factory instance for the model.
     *
     * @param  mixed  $parameters
     * @return \Illuminate\Database\Eloquent\Factories\Factory
     */
    public static function factory(...$parameters)
    {
        $factory = static::newFactory() ?: Factory::factoryForModel(get_called_class());

        return $factory
                    ->count(is_numeric($parameters[0] ?? null) ? $parameters[0] : null)
                    ->state(is_array($parameters[0] ?? null) ? $parameters[0] : ($parameters[1] ?? []));
    }

  // ...
}
```

I tried to "overwrite" the spread `$parameters` to determine the return type, but unfortunately Psalm crashed as a consequence.

### PR DESCRIPTION
**Pull request by caugner**
Fixes #6050.

### CODE REVIEW

**Comment by caugner** [comment]
@weirdan Is it an acceptable approach to have those methods `setParams()` and `addParam()` on the `FunctionLikeStorage`? Maybe I should mark them as `@internal`?

**Review by weirdan — CHANGES_REQUESTED** [review]
> Maybe I should mark them as @internal?

Yes, please do.

File: src/Psalm/Storage/FunctionLikeStorage.php [hunk 1]
```diff
@@ -271,4 +275,20 @@ function (FunctionLikeParameter $param) use ($newlines) : string {
 
         return $visibility_text . ' ' . $symbol_text;
     }
+
+    /**
+     * @param list<FunctionLikeParameter> $params
```
Comment (caugner) [inline src/Psalm/Storage/FunctionLikeStorage.php:280]:
```suggestion
     * @internal
     *
     * @param list<FunctionLikeParameter> $params
```

File: src/Psalm/Storage/FunctionLikeStorage.php [hunk 2]
```diff
@@ -271,4 +275,20 @@ function (FunctionLikeParameter $param) use ($newlines) : string {
...
+     */
+    public function setParams(array $params): void
+    {
+        $this->params = $params;
+        $param_names = array_column($params, 'name');
+        $this->param_lookup = array_fill_keys($param_names, true);
+    }
+
```
Comment (caugner) [inline src/Psalm/Storage/FunctionLikeStorage.php:288]:
```suggestion

    /**
     * @internal
     */
```

**Review by weirdan — APPROVED** [review]

**Comment by weirdan** [comment]
Thanks!

**Pull request merged**